go-controller/pkg/clustermanager/zone_cluster_controller.go

```diff
@@ -0,0 +1,354 @@
+	}
+
+	// Reserve the id 0. We don't want to assign this id to any of the nodes.
+	if err := idAllocator.ReserveID("zero", 0); err != nil {
+		return nil, fmt.Errorf("idAllocator failed to reserve id 0")
+	}
+
+	kube := &kube.Kube{
+		KClient: ovnClient.KubeClient,
+	}
+	wg := &sync.WaitGroup{}
+	zcc := &zoneClusterController{
+		kube:         kube,
+		watchFactory: wf,
+		stopChan:     make(chan struct{}),
+		wg:           wg,
+		idAllocator:  idAllocator,
+	}
+
+	zcc.initRetryFramework()
+	return zcc, nil
+}
+
+func (zcc *zoneClusterController) initRetryFramework() {
+	// We are interested in only nodes
+	resourceHandler := &objretry.ResourceHandler{
+		HasUpdateFunc:          true,
+		NeedsUpdateDuringRetry: false,
+		ObjType:                factory.NodeType,
+		EventHandler: &zoneClusterControllerEventHandler{
+			objType:  factory.NodeType,
+			zcc:      zcc,
+			syncFunc: nil,
+		},
+	}
+
+	zcc.retryNodes = objretry.NewRetryFramework(zcc.stopChan, zcc.wg, zcc.watchFactory, resourceHandler)
+}
+
+// Start starts the zone cluster controller to watch the kubernetes nodes
+func (zcc *zoneClusterController) Start(ctx context.Context) error {
+	nodeHandler, err := zcc.retryNodes.WatchResource()
+
+	if err != nil {
+		return fmt.Errorf("unable to watch nodes: %w", err)
+	}
+
+	zcc.nodeHandler = nodeHandler
+	return nil
+}
+
+func (zcc *zoneClusterController) Stop() {
+	close(zcc.stopChan)
+	zcc.wg.Wait()
+
+	if zcc.nodeHandler != nil {
+		zcc.watchFactory.RemoveNodeHandler(zcc.nodeHandler)
+	}
+}
+
+// handleAddUpdateNodeEvent handles the add or update node event
+func (zcc *zoneClusterController) handleAddUpdateNodeEvent(node *corev1.Node) error {
+	allocatedNodeID, err := zcc.idAllocator.AllocateID(node.Name)
+	if err != nil {
+		return fmt.Errorf("failed to allocate an id to the node %s : err - %v", node.Name, err)
+	}
+	klog.V(5).Infof("Allocated id %d to the node %s", allocatedNodeID, node.Name)
+	nodeAnnotations := util.UpdateNodeIDAnnotation(nil, allocatedNodeID)
+
+	return zcc.updateNodeAnnotationWithRetry(node.Name, nodeAnnotations)
+}
+
+// handleAddUpdateNodeEvent handles the delete node event
+func (zcc *zoneClusterController) handleDeleteNode(node *corev1.Node) error {
+	zcc.idAllocator.ReleaseID(node.Name)
+	return nil
+}
+
+func (zcc *zoneClusterController) syncNodes(nodes []interface{}) error {
+	return zcc.syncNodeIDs(nodes)
+}
+
+func (zcc *zoneClusterController) syncNodeIDs(nodes []interface{}) error {
+	duplicateIdNodes := []string{}
+
+	for _, nodeObj := range nodes {
+		node, ok := nodeObj.(*corev1.Node)
+		if !ok {
+			return fmt.Errorf("spurious object in syncNodes: %v", nodeObj)
+		}
+
+		nodeID := util.GetNodeID(node)
+		if nodeID != -1 {
+			klog.V(5).Infof("Node %s has the id %d set", node.Name, nodeID)
+			if err := zcc.idAllocator.ReserveID(node.Name, nodeID); err != nil {
+				// The id set on this node is duplicate.
+				klog.V(5).Infof("Node %s has a duplicate id [%d] set", node.Name, nodeID)
```

Comment:
This could happen if a crazy person like me,  stops cluster manager and in the meantime runs this command : 

oc annotate node/ovn-worker ovn-node-id=<ovn_worker2_id>

There's a test case which does this.